plugins/anomaly-detection/AnomalyDetection.py

```diff
@@ -0,0 +1,594 @@
+
+            anomaly_result.append(result)
+
+    return anomaly_result
+
+
+def calculate_dbscan(diff_secs_throughput):
+    """
+    The function calculates Density-based spatial clustering of applications
+    with Noise (DBSCAN) for a given list of throughput values of a connection
+    Args:
+        dataframe: The row of a dataframe containing all data related to this
+        network connection.
+        Assumption: Since DBSCAN needs only numeric value to train and start
+        prediction, any connection with null values will not be taken
+        into account for calculation. We return empty value in that case.
+    Returns:
+        A list of boolean values which signifies if a network flow records
+        of the connection is anomalous or not based on DBSCAN
+    """
+    # Currently just a placeholder function
+    placeholder_throughput_list = []
+    for i in range(len(diff_secs_throughput)):
+        placeholder_throughput_list.append(0.0)
+    return placeholder_throughput_list
+
+
+def calculate_dbscan_anomaly(dataframe):
+    """
+    The function calculates Density-based spatial clustering of applications
+    with Noise (DBSCAN) for a given list of throughput values of a connection
+    Args:
+        dataframe: The row of a dataframe containing all data related to this
+        network connection.
+        Assumption: Since DBSCAN needs only numeric value to train and start
+        prediction, any connection with null values will not be taken
+        into account for calculation. We return empty value in that case.
+    Returns:
+        A list of boolean values which signifies if a network flow records
+        of the connection is anomalous or not based on DBSCAN
+    """
+
+    throughput_list = dataframe[10]
+    anomaly_result = []
+
+    np_throughput_list = np.array(throughput_list)
+    np_throughput_list = np_throughput_list.reshape(-1, 1)
+
+    outlier_detection = DBSCAN(min_samples=4, eps=250000000)
+    clusters = outlier_detection.fit_predict(np_throughput_list)
+
+    for i in clusters:
+        if i == -1:
+            anomaly_result.append(True)
+        else:
+            anomaly_result.append(False)
+
+    return anomaly_result
+
+def filter_df_with_true_anomalies(spark, plotDF, algo_type):
+    plotDF = plotDF.withColumn(
+        "new", f.arrays_zip("flowEndSeconds", "algoCalc", "Throughputs", "anomaly")).withColumn(
+        "new", f.explode("new")).select(
+            "sourceIP", "sourceTransportPort", "destinationIP", "destinationTransportPort",
+            "protocolIdentifier", "flowStartSeconds", f.col("new.flowEndSeconds").alias("flowEndSeconds"),
+             "throughputStandardDeviation", "algoType", f.col("new.algoCalc").alias("algoCalc"),
+             f.col("new.Throughputs").alias("Throughputs"), f.col("new.anomaly").alias("anomaly"))
+    ret_plot = plotDF.where(plotDF.anomaly!=False)
+    if ret_plot.count()==0:
+        ret_plot = ret_plot.collect()
```

Comment:
May I ask why we use `ret_plot.collect()` here? Since `ret_plot.count()==0`, I think we could initialize `ret_plot` with an empty list

Reply:
yes, it is not required, left from the testing phase, removing it